File: AnalyticsKitUnitTestProvider.m
//
//  AnalyticsKitUnitTestProvider.m
//  TeamStream
//
//  Created by Todd Huss on 11/14/12.
//  Copyright (c) 2012 Two Bit Labs. All rights reserved.
//

#import "AnalyticsKitUnitTestProvider.h"

@implementation AnalyticsKitUnitTestProvider

+ (AnalyticsKitUnitTestProvider *)setUp {
    AnalyticsKitUnitTestProvider *provider = [[AnalyticsKitUnitTestProvider alloc] init];
    provider.events = [NSMutableArray arrayWithCapacity:20];
    [AnalyticsKit initializeLoggers:@[provider]];
    return provider;
}

+ (AnalyticsKitUnitTestProvider *)unitTestProvider {
    AnalyticsKitUnitTestProvider *unitProvider = nil;
    NSArray *loggers = [AnalyticsKit loggers];
    for (AnalyticsKitUnitTestProvider *provider in loggers) {
        if ([provider isKindOfClass:[AnalyticsKitUnitTestProvider class]]) unitProvider = provider;
    }
    return unitProvider;
}

+ (void)clearEvents {
    // Remove the events stored in the unit test provider
    [[self unitTestProvider] clearEvents];
}

+ (void)tearDown {
    [self clearEvents];
    // Wipe out any loggers
    [AnalyticsKit initializeLoggers:[[NSArray alloc] init]];
}

- (void)clearEvents {
    [self setEvents:[NSMutableArray arrayWithCapacity:20]];
}

- (BOOL)hasEventLoggedWithName:(NSString *)eventName {
    return [self firstEventLoggedWithName:eventName] != nil;
}

- (AnalyticsKitEvent *)firstEventLoggedWithName:(NSString *)eventName {
    AnalyticsKitEvent *event = nil;
    NSArray *matchingEvents = [self eventsLoggedWithName:eventName];
    if ([matchingEvents count] > 0) event = matchingEvents[0];
    return event;
}

- (NSArray *)eventsLoggedWithName:(NSString *)eventName {
    NSMutableArray *matchingEvents = [NSMutableArray arrayWithCapacity:5];
    for (AnalyticsKitEvent *event in self.events) {
        if ([eventName isEqualToString:event.name]) [matchingEvents addObject:event];
    }
    return matchingEvents;
}

#pragma mark -
#pragma mark Lifecycle

-(void)applicationWillEnterForeground{}

-(void)applicationDidEnterBackground{}

-(void)applicationWillTerminate{}

-(void)uncaughtException:(NSException *)exception{}

#pragma mark -
#pragma mark Event Logging

-(void)logScreen:(NSString *)screenName{
    NSString *event = [@"Screen - " stringByAppendingString:screenName];
    [self logEvent:event withProperties:nil];
}

-(void)logEvent:(NSString *)event {
    [self logEvent:event withProperties:nil];
}

-(void)logEvent:(NSString *)event withProperty:(NSString *)key andValue:(NSString *)value {
    [self logEvent:event withProperties:@{key:value}];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict {
    [self.events addObject:[[AnalyticsKitEvent alloc] initEvent:event withProperties:dict]];
}

-(void)logEvent:(NSString *)event timed:(BOOL)timed{
    [self logEvent:event withProperties:nil];
}

-(void)logEvent:(NSString *)event withProperties:(NSDictionary *)dict timed:(BOOL)timed{
    [self logEvent:event withProperties:dict];
}

-(void)endTimedEvent:(NSString *)event withProperties:(NSDictionary *)dict{}

-(void)logError:(NSString *)name message:(NSString *)message exception:(NSException *)exception{}

-(void)logError:(NSString *)name message:(NSString *)message error:(NSError *)error{}

@end

File: AnalyticsKit/AnalyticsKit/AnalyticsKitUnitTestProvider.swift
import Foundation

public class AnalyticsKitUnitTestProvider: NSObject, AnalyticsKitProvider {

    var events = [AnalyticsKitEvent]()

    public class func setUp() -> AnalyticsKitUnitTestProvider {
        let provider: AnalyticsKitUnitTestProvider = AnalyticsKitUnitTestProvider()
        provider.events = [AnalyticsKitEvent]()
        AnalyticsKit.initializeProviders([provider])
        return provider
    }

    public class func unitTestProvider() -> AnalyticsKitUnitTestProvider? {
        var unitProvider: AnalyticsKitUnitTestProvider?
        for provider: AnalyticsKitProvider in AnalyticsKit.providers() {
            if (provider is AnalyticsKitUnitTestProvider) {
                unitProvider = provider as? AnalyticsKitUnitTestProvider
            }
        }
        return unitProvider
    }

    public class func clearEvents() {
        // Remove the events stored in the unit test provider
        self.unitTestProvider()?.clearEvents()
    }

    public class func tearDown() {
        self.clearEvents()
        // Wipe out any loggers
        AnalyticsKit.initializeProviders([AnalyticsKitProvider]())
    }


    func clearEvents() {
        self.events = [AnalyticsKitEvent]()
    }

    public func hasEventLoggedWithName(_ eventName: String) -> Bool {
        return firstEventLoggedWithName(eventName) != nil
    }

    public func firstEventLoggedWithName(_ eventName: String) -> AnalyticsKitEvent? {
        var event: AnalyticsKitEvent?
        var matchingEvents: [AnalyticsKitEvent] = eventsLoggedWithName(eventName)
        if matchingEvents.count > 0 {
            event = matchingEvents[0]
        }
        return event
    }

    public func eventsLoggedWithName(_ eventName: String) -> [AnalyticsKitEvent] {
        var matchingEvents: [AnalyticsKitEvent] = [AnalyticsKitEvent]()
        for event in events {
            if (eventName == event.name) {
                matchingEvents.append(event)
            }
        }
        return matchingEvents
    }

    public func applicationWillEnterForeground() {}
    public func applicationDidEnterBackground() {}
    public func applicationWillTerminate() {}
    public func uncaughtException(_ exception: NSException) {}
    public func endTimedEvent(_ event: String, withProperties properties: [String: Any]) {}
    public func logError(_ name: String, message: String?, properties: [String: Any]?, exception: NSException?) {}
    public func logError(_ name: String, message: String?, properties: [String: Any]?, error: Error?) {}

    public func logScreen(_ screenName: String) {
        logEvent("Screen - \(screenName)")
    }

    public func logScreen(_ screenName: String, withProperties properties: [String: Any]) {
        logEvent("Screen - \(screenName)", withProperties: properties)
    }

    public func logEvent(_ event: String) {
        self.events.append(AnalyticsKitEvent(event: event))
    }

    public func logEvent(_ event: String, withProperty key: String, andValue value: String) {
        self.logEvent(event, withProperties: [key: value])
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any]) {
        self.events.append(AnalyticsKitEvent(event: event, withProperties: properties))
    }

    public func logEvent(_ event: String, timed: Bool) {
        self.events.append(AnalyticsKitEvent(event: event))
    }

    public func logEvent(_ event: String, withProperties properties: [String: Any], timed: Bool) {
        self.logEvent(event, withProperties: properties)
    }
}
